crates/accelerate/src/two_qubit_decompose.rs

```diff
@@ -180,20 +272,919 @@ fn __num_basis_gates(basis_b: f64, basis_fidelity: f64, unitary: MatRef<c64>) ->
+            Specializations::PartialSWAPEquiv => 3,
+            Specializations::PartialSWAPFlipEquiv => 4,
+            Specializations::ControlledEquiv => 5,
+            Specializations::MirrorControlledEquiv => 6,
+            Specializations::SimaabEquiv => 7,
+            Specializations::SimabbEquiv => 8,
+            Specializations::SimabmbEquiv => 9,
+        };
+        (
+            [
+                self.a,
+                self.b,
+                self.c,
+                self.global_phase,
+                self.calculated_fidelity,
+            ],
+            [
+                self.K1l.to_pyarray(py).into(),
+                self.K1r.to_pyarray(py).into(),
+                self.K2l.to_pyarray(py).into(),
+                self.K2r.to_pyarray(py).into(),
+                self.unitary_matrix.to_pyarray(py).into(),
+            ],
+            specialization,
+            self.default_euler_basis.clone(),
+            self.requested_fidelity,
+        )
+    }
+
+    fn __setstate__(
+        &mut self,
+        state: (
+            [f64; 5],
+            [PyReadonlyArray2<Complex64>; 5],
+            u8,
+            String,
+            Option<f64>,
+        ),
+    ) {
+        self.a = state.0[0];
+        self.b = state.0[1];
+        self.c = state.0[2];
+        self.global_phase = state.0[3];
+        self.calculated_fidelity = state.0[4];
+        self.K1l = state.1[0].as_array().to_owned();
+        self.K1r = state.1[1].as_array().to_owned();
+        self.K2l = state.1[2].as_array().to_owned();
+        self.K2r = state.1[3].as_array().to_owned();
+        self.unitary_matrix = state.1[4].as_array().to_owned();
+        self.default_euler_basis = state.3;
+        self.requested_fidelity = state.4;
+        self.specialization = match state.2 {
+            0 => Specializations::General,
+            1 => Specializations::IdEquiv,
+            2 => Specializations::SWAPEquiv,
+            3 => Specializations::PartialSWAPEquiv,
+            4 => Specializations::PartialSWAPFlipEquiv,
+            5 => Specializations::ControlledEquiv,
+            6 => Specializations::MirrorControlledEquiv,
+            7 => Specializations::SimaabEquiv,
+            8 => Specializations::SimabbEquiv,
+            9 => Specializations::SimabmbEquiv,
+            _ => unreachable!("Invalid specialization value"),
+        };
+    }
+
+    fn __getnewargs__(&self, py: Python) -> (PyObject, Option<f64>, Option<Specializations>, bool) {
+        (
+            self.unitary_matrix.to_pyarray(py).into(),
+            self.requested_fidelity,
+            None,
+            true,
+        )
+    }
+
+    #[new]
+    #[pyo3(signature=(unitary_matrix, fidelity=DEFAULT_FIDELITY, _specialization=None, _pickle_context=false))]
+    fn new(
+        unitary_matrix: PyReadonlyArray2<Complex64>,
+        fidelity: Option<f64>,
+        _specialization: Option<Specializations>,
+        _pickle_context: bool,
+    ) -> PyResult<Self> {
```

Comment:
Moving this down to Rust-space has meant (evidenced by changes in the test suite) that the public interface to the constructor has changed; now it's only objects that are already interpretable as `PyReadonlyArray2`, whereas before it accepted anything that Numpy could coerce to an array.

I think that's a sufficiently large breaking change that we should take care of it properly - accept `Py<PyAny>` in the constructor, and do the extraction ourselves in Rust space, or use the the Python inheriting wrapper to call `np.asarray(unitary_matrix, dtype=complex)` before it gets passed to here.

Reply:
That's a good call, we I was thinking of adjusting the python side constructor to do this, but we can easily do this rust side. I guess it really comes down to if we refactor the to just wrap an inner rust private method per: https://github.com/Qiskit/qiskit/pull/11946#discussion_r1520192861 or not. If we do that we should just do this python side.

Reply:
Fixed in: https://github.com/Qiskit/qiskit/pull/11946/commits/a2fcc58a9edd2c1b1253d88a33d8054d0eca47f7